internal/smartlink/smartlink.go

```diff
@@ -0,0 +1,143 @@
+package smartlink
+
+import (
+	"os"
+	"path/filepath"
+
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/fileutils"
+	"github.com/ActiveState/cli/internal/logging"
+)
+
+// LinkContents will link the contents of src to desc
+func LinkContents(src, dest string) error {
+	if !fileutils.DirExists(src) {
+		return errs.New("src dir does not exist: %s", src)
+	}
+	if err := fileutils.MkdirUnlessExists(dest); err != nil {
+		return errs.Wrap(err, "Could not create dir: %s", dest)
+	}
+
+	var err error
+	src, dest, err = resolvePaths(src, dest)
+	if err != nil {
+		return errs.Wrap(err, "Could not resolve src and dest paths")
+	}
+
+	entries, err := os.ReadDir(src)
+	if err != nil {
+		return errs.Wrap(err, "Reading dir %s failed", src)
+	}
+	for _, entry := range entries {
+		if err := Link(filepath.Join(src, entry.Name()), filepath.Join(dest, entry.Name())); err != nil {
+			return errs.Wrap(err, "Link failed")
+		}
+	}
+
+	return nil
+}
+
+// Link creates a link from src to target. MS decided to support Symlinks but only if you opt into developer mode (go figure),
+// which we cannot reasonably force on our users. So on Windows we will instead create dirs and hardlinks.
+func Link(src, dest string) error {
+	var err error
+	src, dest, err = resolvePaths(src, dest)
+	if err != nil {
+		return errs.Wrap(err, "Could not resolve src and dest paths")
+	}
+
+	if fileutils.IsDir(src) {
+		if err := os.MkdirAll(dest, 0755); err != nil {
+			return errs.Wrap(err, "could not create directory %s", dest)
+		}
+		entries, err := os.ReadDir(src)
+		if err != nil {
+			return errs.Wrap(err, "could not read directory %s", src)
+		}
+		for _, entry := range entries {
+			if err := Link(filepath.Join(src, entry.Name()), filepath.Join(dest, entry.Name())); err != nil {
+				return errs.Wrap(err, "sub link failed")
+			}
+		}
+		return nil
+	}
+
+	// Multiple artifacts can supply the same file. We do not have a better solution for this at the moment other than
+	// favouring the first one encountered.
+	if fileutils.TargetExists(dest) {
+		logging.Warning("Skipping linking '%s' to '%s' as it already exists", src, dest)
+		return nil
+	}
+
+	if err := linkFile(src, dest); err != nil {
+		return errs.Wrap(err, "could not link %s to %s", src, dest)
+	}
+	return nil
```

Comment:
nit:
```suggestion
	return linkFile(src, dest)
```

But then you may have to adjust the error in the caller.